### PR TITLE
clippy fixes for librustdoc

### PR DESCRIPTION
**Pull request by matthiaskrgr**
fixes clippy warnings of type:
match_like_matches_macro
or_fun_call
op_ref
needless_return
let_and_return
single_char_add_str
useless_format
unnecessary_sort_by
match_ref_pats
redundant_field_names

### CODE REVIEW

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -1641,7 +1641,7 @@ fn report_diagnostic(
     };
 
     let attrs = &item.attrs;
-    let sp = span_of_attrs(attrs).unwrap_or(item.source.span());
+    let sp = span_of_attrs(attrs).unwrap_or_else(|| item.source.span());
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:1644]: span() is trivial.

File: src/librustdoc/passes/doc_test_lints.rs [hunk 2]
```diff
@@ -96,7 +96,7 @@ crate fn look_for_tests<'tcx>(cx: &DocContext<'tcx>, dox: &str, item: &Item) {
     if tests.found_tests == 0 && cx.tcx.sess.is_nightly_build() {
         if should_have_doc_example(cx, &item) {
             debug!("reporting error for {:?} (hir_id={:?})", item, hir_id);
-            let sp = span_of_attrs(&item.attrs).unwrap_or(item.source.span());
+            let sp = span_of_attrs(&item.attrs).unwrap_or_else(|| item.source.span());
```
Comment (jyn514) [inline src/librustdoc/passes/doc_test_lints.rs:99]: etc. for the rest of the span() calls.

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/librustdoc/clean/cfg.rs [hunk 3]
```diff
@@ -177,29 +177,20 @@ impl Cfg {
             Cfg::Any(ref sub_cfgs) | Cfg::All(ref sub_cfgs) => {
                 sub_cfgs.first().map(Cfg::should_capitalize_first_letter).unwrap_or(false)
             }
-            Cfg::Cfg(name, _) => match name {
-                sym::debug_assertions | sym::target_endian => true,
-                _ => false,
-            },
+            Cfg::Cfg(name, _) => matches!(name, sym::debug_assertions | sym::target_endian),
```
[the remark this answers is not in the document]
Comment (matthiaskrgr) [reply src/librustdoc/clean/cfg.rs:180]: `Cfg::Cfg(name, _) => name == sym::debug_assertions || name ==  sym::target_endian,` ?

File: src/librustdoc/clean/types.rs [hunk 4]
```diff
@@ -378,10 +376,7 @@ impl ItemKind {
     }
 
     crate fn is_type_alias(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (matthiaskrgr) [reply src/librustdoc/clean/types.rs:378]: It's used here https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/render/mod.rs#L3843

[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/types.rs:378]: Oh ugh, trait impls are a mess and people complain whenever we change the display. Ok, this is fine for now then.

File: src/librustdoc/clean/cfg.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/cfg.rs:180]: Yup! That looks good :)

File: src/librustdoc/html/format.rs [hunk 5]
```diff
@@ -244,8 +244,8 @@ impl<'a> fmt::Display for WhereClause<'a> {
                 clause.push_str("<br>");
             }
 
-            match pred {
-                &clean::WherePredicate::BoundPredicate { ref ty, ref bounds } => {
+            match *pred {
+                clean::WherePredicate::BoundPredicate { ref ty, ref bounds } => {
```
Comment (LingMan) [inline src/librustdoc/html/format.rs:248]: Unneeded deref and unneeded `ref` keywords. Same for the other arms of this `match`.

**Review by LingMan — CHANGES_REQUESTED** [review]

File: src/librustdoc/clean/cfg.rs [hunk 6]
```diff
@@ -177,29 +177,20 @@ impl Cfg {
...
+            Cfg::Not(ref child) => matches!(**child, Cfg::Cfg(..)),
         }
     }
 
     fn should_use_with_in_description(&self) -> bool {
-        match *self {
-            Cfg::Cfg(name, _) if name == sym::target_feature => true,
-            _ => false,
-        }
+        matches!(*self, Cfg::Cfg(name, _) if name == sym::target_feature)
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/cfg.rs:193]: @LingMan no, `target_feature` is a constant, not an enum variant.

[the remark this answers is not in the document]
Comment (LingMan) [reply src/librustdoc/clean/cfg.rs:193]: @jyn514 Why would that be a problem?
```
fn should_use_with_in_description(&self) -> bool {
    matches!(self , Cfg::Cfg(sym::target_feature, _))
}
```
passes `./x.py check` just fine. Is there a semantic difference I'm not aware of?

[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/cfg.rs:193]: I didn't know that was a valid pattern :eyes: I found that [it works](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2b83c5cd77948578dc1ce82ee1f6908f) even if there's [no path in front](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c9c88098a92c53c04540e9c658213348), but I could have sworn that meant a variable binding instead ... I'm really confused :confused:

[the remark this answers is not in the document]
Comment (LingMan) [reply src/librustdoc/clean/cfg.rs:193]: It becomes a variable binding if you remove the `use inner::C;`.

File: src/librustdoc/clean/cfg.rs [hunk 7]
```diff
@@ -177,29 +177,20 @@ impl Cfg {
...
 
     fn should_append_only_to_description(&self) -> bool {
         match *self {
             Cfg::False | Cfg::True => false,
             Cfg::Any(..) | Cfg::All(..) | Cfg::Cfg(..) => true,
-            Cfg::Not(ref child) => match **child {
-                Cfg::Cfg(..) => true,
-                _ => false,
-            },
+            Cfg::Not(ref child) => matches!(**child, Cfg::Cfg(..)),
```
[the remark this answers is not in the document]
Comment (matthiaskrgr) [reply src/librustdoc/clean/cfg.rs:188]: Cfg::Not contains a `Box<Cfg>`, I'm not sure how to match on that.

[the remark this answers is not in the document]
Comment (LingMan) [reply src/librustdoc/clean/cfg.rs:188]: `Cfg::Not(box Cfg::Cfg(..))`

[the remark this answers is not in the document]
Comment (matthiaskrgr) [reply src/librustdoc/clean/cfg.rs]: interesting, thanks :D

**Pull request merged**